### PR TITLE
Hierarchical state machines - part 2

### PR DESCRIPTION
**Pull request by bergundy**
"Backend" of the HSM framework including changes to mutable state, task definitions, serializers, generator, and executors.

Note that mutable state itself is registered as a state machine in the HSM framework and used as the root node in the HSM tree.

It's a bit annoying that all of the unit tests need to register the workflow state machine but I think this is ended up being the cleanest approach from the HSM framework perspective.

There's still some follow-up work, like collapsing timer tasks and support for interacting with history event, that I'll take on later.

Replication support for the framework will also come at a later time.

### CODE REVIEW

File: service/history/fx.go [hunk 1]
```diff
@@ -84,8 +91,22 @@ var Module = fx.Options(
 	fx.Provide(HandlerProvider),
 	fx.Provide(ServiceProvider),
 	fx.Invoke(ServiceLifetimeHooks),
+
+	fx.Provide(CallbackExecutorOptionsProvider),
+	callbacks.Module,
 )
 
+func CallbackExecutorOptionsProvider() callbacks.ActiveExecutorOptions {
```
Comment (bergundy) [inline service/history/fx.go:99]: I'm debating moving this into the callbacks plugin. This is a dependency of the callback executor.

File: service/history/queues/executable.go [hunk 2]
```diff
@@ -155,7 +155,7 @@ func NewStateStaleError(message string) StaleStateError {
 }
 
 func (e StaleStateError) Error() string {
-	return "task stale: " + e.Message
+	return "state stale: " + e.Message
```
Comment (bergundy) [inline service/history/queues/executable.go:158]: Glad I caught this before it was deployed, could have made for some confusing debugging.

File: service/history/shard/context_testutil.go [hunk 3]
```diff
@@ -206,6 +207,11 @@ func (s *ContextTest) SetHistoryClientForTesting(client historyservice.HistorySe
 	s.historyClient = client
 }
 
+// SetStateMachineRegistry the state machine registry on this shard.
```
Comment (bergundy) [inline service/history/shard/context_testutil.go:210]:
```suggestion
// SetStateMachineRegistry sets the state machine registry on this shard.
```

File: service/history/workflow/mutable_state_impl.go [hunk 4]
```diff
@@ -3825,10 +3873,20 @@ func (ms *MutableStateImpl) AddContinueAsNewEvent(
 	); err != nil {
 		return nil, nil, err
 	}
-	// Copy over close callbacks to the next execution.
-	for id, cb := range ms.executionInfo.Callbacks {
+
+	// TODO: should this be in the "Apply" function? How does this work with replication?
+	// TODO: also copy over close callbacks if workflow completes unsuccessfully and its retry policy permits it.
```
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:3878]: I'll address these post refactor.

File: service/history/workflow/mutable_state.go [hunk 5]
```diff
@@ -70,6 +72,31 @@ func (policy TransactionPolicy) Ptr() *TransactionPolicy {
 
 var emptyTasks = []tasks.Task{}
 
+// Mutable state is a top-level state machine in the state machines framework.
+var StateMachineType = hsm.MachineType{
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/workflow/mutable_state.go:76]: I want all of the state machines used in the HSM framework to have a definition and not have to make exceptions there.

We can add more metadata to the definition that may be useful later.

File: service/history/workflow/mutable_state.go [hunk 6]
```diff
@@ -110,7 +137,7 @@ type (
 	}
 
 	MutableState interface {
-		statemachines.Environment
+		callbacks.CanGetNexusCompletion
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/workflow/mutable_state.go:140]: Wondering what makes you say that.

File: service/history/workflow/mutable_state_impl.go [hunk 7]
```diff
@@ -4705,16 +4788,29 @@ func (ms *MutableStateImpl) prepareCloseTransaction(
 		return err
 	}
 
+	ms.executionInfo.StateTransitionCount += 1
+
+	ms.executionInfo.TransitionHistory = UpdatedTransitionHistory(
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go:4793]: Yeah, we should handle this as part of the replication work. (@MichaelSnowden heads up)

I'll track.

File: service/history/workflow/state_transition_history.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+// TransitionHistoryStalenessCheck verifies that task namespace failover version and transition count is contained in
+// the given transition history.
+//
+// When a task is being processed, the history is compared with the imprinted task information to verify that a task is
+// not referencing a stale state or that the task itself is not stale. For example, if the state has a history of
+// `[{v: 1, t: 3}, {v: 2, t: 5}]`, task A `{v: 2, t: 4}` **is not** referencing stale state because for version `2`
+// transitions `4-5` are valid, while task B `{v: 2, t: 6}` **is** referencing stale state because the transition count
+// is out of range for version `2`. Furthermore, task C `{v: 1, t: 4}` itself is stale because it is referencing an
+// impossible state, likely due to post split-brain reconciliation.
+func TransitionHistoryStalenessCheck(history []*persistencespb.VersionedTransition, taskNamespaceFailoverVersion, taskStateTransitionCount int64) error {
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/workflow/state_transition_history.go:68]: Yes, I figured but I wanted to have meaningful errors for task staleness checks.

File: common/persistence/serialization/task_serializer.go [hunk 9]
```diff
@@ -1149,79 +1176,42 @@ func (s *TaskSerializer) replicationSyncWorkflowStateTaskFromProto(
...
+		return nil, serviceerror.NewInternal(fmt.Sprintf("unknown callback task type while deserializing: %v", info))
 	}
+	return &tasks.StateMachineCallbackTask{
+		StateMachineTask: tasks.StateMachineTask{
+			WorkflowKey: definition.NewWorkflowKey(
+				info.NamespaceId,
+				info.WorkflowId,
+				info.RunId,
+			),
+			VisibilityTimestamp: time.Unix(0, 0).UTC(),
```
[the remark this answers is not in the document]
Comment (bergundy) [reply common/persistence/serialization/task_serializer.go:1211]: Thanks, I'll address.

File: service/history/shard/context_impl.go [hunk 10]
```diff
@@ -2183,6 +2188,10 @@ func (s *ContextImpl) GetArchivalMetadata() archiver.ArchivalMetadata {
 	return s.archivalMetadata
 }
 
+func (s *ContextImpl) StateMachineRegistry() *hsm.Registry {
+	return s.stateMachineRegistry
+}
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/shard/context_impl.go:2193]: I need this in mutable state too. I think we'll end up needing it in a bunch of places.

**Review by bergundy — COMMENTED** [review]
Thanks for the review.
I'll push a new commit shortly.

File: service/history/callback_queue_factory.go [hunk 11]
```diff
@@ -93,16 +63,14 @@ var _ ctasks.DynamicWorkerPoolLimiter = groupLimiter{}
...
-	// Shared client pool for all shards in the host.
-	completionCallerPool *groupPool[queues.NamespaceIDAndDestination, HTTPCaller]
 	// Shared rate limiter pool for all shards in the host.
-	rateLimiterPool *groupPool[queues.NamespaceIDAndDestination, quotas.RateLimiter]
+	rateLimiterPool *collection.OnceMap[queues.NamespaceIDAndDestination, quotas.RateLimiter]
 	// Shared scheduler across all shards in the host.
 	hostScheduler queues.Scheduler
 }
 
 func NewCallbackQueueFactory(params callbackQueueFactoryParams) QueueFactory {
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/callback_queue_factory.go:72]: Yes, I'll refactor after this PR.
It's already too big.
I'm trying to contain it just to the state machine "backend".

File: service/history/shard/context_impl.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/shard/context_impl.go:2193]: ah I see. mutable state... yeah then shardContext is probably the best place...

File: service/history/task_executor.go [hunk 12]
```diff
@@ -208,3 +177,44 @@ func (l *taskExecutor) getValidatedMutableStateForTask(
...
+		return err
+	}
+	if accessType == hsm.AccessRead {
+		return nil
+	}
+
+	if ms.GetExecutionState().State == enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED {
+		// Can't use UpdateWorkflowExecutionAsActive since it updates the current run, and we are operating on closed
+		// workflows.
+		return wfCtx.SubmitClosedWorkflowSnapshot(ctx, t.shardContext, workflow.TransactionPolicyActive)
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/task_executor.go:210]: hmm, no... for cloud should be fine reusing the existing impl.

File: service/history/workflow/mutable_state.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/workflow/mutable_state.go:76]: sure. I guess I was thinking the framework doesn't quite support top level state machine now :) I don't have objection defining it. We will define it sooner or later.

File: service/history/workflow/mutable_state.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/workflow/mutable_state.go:140]: I was thinking `MutableState` is a framework level implementation and any workflow specific logic should be split out to a separate state machine implementation. So ideally it should not be aware of state machine specific logic/definition. But yeah right now those two are the same object.

File: service/history/workflow/mutable_state_impl.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/workflow/mutable_state_impl.go:4793]: I am ok with landing this. But if this implementation ever get released, we will likely need a new field as the history computed by the existing implementation doesn't look right to me under replication/reset case.

**Review by yycptt — APPROVED** [review]

File: service/history/workflow/mutable_state_impl.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go:4793]: 100% agree.

**Pull request merged**